Add GitHub Actions

luisherranz opened the pull request:
<!--
Thanks for your pull request 😊. Note that not following the template might result in your issue being closed
-->

#### Description of what you did:

In this PR I'm going to experiment with GH actions. If I'm able to make it work, we'll switch from Travis.

#### My PR is a:

<!-- Delete the ones that don't apply -->

- 🔝 CI Improvement

#### Is the PR ready to be merged?

<!-- Delete the one that don't apply -->

- Yes!

luisherranz committed:
Add unit-tests workflow to GH actions

Very first version. I'm just testing GH actions.

luisherranz committed:
Delete .travis.yml

luisherranz committed:
Add e2e using cypress action

Let's try first with the basic example of https://github.com/cypress-io/github-action

luisherranz committed:
Rename unit-tests job

luisherranz committed:
Rename e2e-tests job

luisherranz committed:
Rename unit tests

luisherranz committed:
Start e2e project before e2e tests

luisherranz committed:
Use npm script to start server

luisherranz committed:
Add longer timeout to wait-on

luisherranz committed:
Add HEADLESS env variable

This env variable is used to avoid tests that are only valid for the "headed"  environment (current version of Chrome).

luisherranz committed:
Add cache to npm in unit tests

luisherranz committed:
Add missing "run" to npm script

luisherranz committed:
Rename e2e step

luisherranz committed:
Refactor unit tests workflow steps

luisherranz committed:
Add parallelization and recording

luisherranz committed:
Re-run checks

luisherranz committed:
Change cypress group name

luisherranz committed:
Add lighthouse to deploy in now

luisherranz committed:
Add now token env variable

luisherranz committed:
Add lighthouse action

luisherranz committed:
Lighthouse step needs deploy

luisherranz committed:
Upload Lighthouse results to public storage

luisherranz committed:
Move LHCI token to env variable

luisherranz committed:
Fix yaml indentation

luisherranz committed:
Simplify lighthouse job to see if it succeeds

luisherranz committed:
Add commit sha to env for lighthouse

luisherranz committed:
Add GitHub token to LHCI

luisherranz committed:
Run LH report 3 times

So the Now CDN can cache the result.

luisherranz committed:
Move e2e to its own folder and add es5 and module

By moving cypress to its own folder we avoid having to reinstall it on
each other GitHub action.

By adding es5 and module separately, we can test those versions as well.

luisherranz committed:
Move e2e project scripts to the root

Because the GH action needs them there.

luisherranz committed:
Change test:clear for test:ci in unit tests

luisherranz committed:
Remove cypress record but upload artifacts

Because we are already over the limit of the free plan of cypress.

luisherranz committed:
Add video and screenshot folders and record videos

Because if folder does not exist, the GH action to upload the
artifacts fails.

luisherranz committed:
Turn off videos and screenshots locally

luisherranz committed:
Upload LH results as artifacts

luisherranz committed:
Add performance assertions in LH

luisherranz committed:
Remove tests on push

Because now with GH actions everything gets properly tested in the PR

luisherranz committed:
Add missing checkout on LH job

luisherranz committed:
Add ci.assert wrapper to the LH config file

luisherranz committed:
Fix LH assert

I made it fail on purpose to check that it worked.

luisherranz committed:
Add budget to LH

luisherranz committed:
Fix LH budget

I made it fail on purpose to check that it worked.

DAreRodz commented on .github/workflows/e2e-tests-all.yml:
```diff
@@ -0,0 +1,48 @@
+      - name: Run e2e tests (all)
+        uses: cypress-io/github-action@v1
+        with:
+          start: npm run e2e:prod
```
The three `e2e` test files are identical except for this line. Is it possible to use a single file and just change the command that will be executed? Maybe using 

```yaml
strategy:
  matrix:
    test-command: [prod, prod:module, prod:es5]
```
and then, in `steps`

```yaml
- name: Run e2e tests (${{ matrix.test-command }})
  uses: cypress-io/github-action@v1
  with:
    start: npm run e2e:${{ matrix.test-command }}
```

like in `unit-tests`, or something like that?

I know nothing about GitHub Actions, BTW. 🙈

luisherranz replied:
Maybe it can be done with `ci_node_total` and `ci_node_index` like [this guy does for cypress](https://gist.github.com/ArturT/3535fc24bf2fac577462b9162c12c1d6), but it is not documented anywhere. I'll give it a try.

luisherranz replied:
It worked using `ci_node_total` and `ci_node_index` https://github.com/frontity/frontity/pull/244/commits/f4b10d0d300fc51e0c5e436339cff2b05036bf5f :)

Thanks for the suggestion @DAreRodz!

DAreRodz reviewed (approved):
Great job.⭐
I have a question about `e2e` files, tho. Let me know what you think.

segunadebayo reviewed (approved):
I think this looks good. Great job, Luis 💫.

luisherranz committed:
Try only one file for e2e and fix video uploads

luisherranz committed:
Remove fail-fast

It was there because I copied the strategy from
https://gist.github.com/ArturT/3535fc24bf2fac577462b9162c12c1d6
not because it was needed.
